File: coding_theory/q6_locating_dominating_16/verify_q6_ld16.py
#!/usr/bin/env python3
"""Solver-free certificate checker for gamma^LD(Q_6) = 16."""

from __future__ import annotations

import argparse
from collections import Counter
from fractions import Fraction


DIMENSION = 6
EXPECTED_WORDS = {
    "000000", "000101", "001011", "001110",
    "010010", "010111", "011001", "011100",
    "100010", "100100", "101001", "101111",
    "110011", "110101", "111000", "111110",
}


def algebraic_code() -> set[int]:
    """Return C, encoding the displayed word x1...x6 as a binary integer."""
    code: set[int] = set()
    for x1 in (0, 1):
        for x4 in (0, 1):
            for x5 in (0, 1):
                for x6 in (0, 1):
                    x2 = x1 ^ x4 ^ x5 ^ x6
                    x3 = x1 ^ x4 ^ x6 ^ (x1 & x5) ^ (x1 & x6)
                    word = (x1, x2, x3, x4, x5, x6)
                    code.add(int("".join(map(str, word)), 2))
    return code


def closed_neighborhood(vertex: int) -> set[int]:
    return {vertex, *(vertex ^ (1 << coordinate) for coordinate in range(DIMENSION))}


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--show-signatures", action="store_true")
    args = parser.parse_args()

    code = algebraic_code()
    expected = {int(word, 2) for word in EXPECTED_WORDS}
    assert code == expected
    assert len(code) == 16

    signatures: dict[frozenset[int], int] = {}
    sizes: Counter[int] = Counter()
    for vertex in range(1 << DIMENSION):
        if vertex in code:
            continue
        signature = frozenset(code & closed_neighborhood(vertex))
        assert signature, f"undominated non-codeword: {vertex:06b}"
        assert signature not in signatures, (
            f"equal signatures at {signatures.get(signature):06b} and {vertex:06b}"
        )
        signatures[signature] = vertex
        sizes[len(signature)] += 1
        if args.show_signatures:
            words = " ".join(f"{word:06b}" for word in sorted(signature))
            print(f"{vertex:06b}: {words}")

    assert len(signatures) == 48
    assert sizes == Counter({1: 16, 2: 16, 3: 16})

    n = DIMENSION
    lower_bound = Fraction(n * n * (1 << (n + 1)), n**3 + 2 * n * n + 3 * n - 2)
    lower_ceiling = (lower_bound.numerator + lower_bound.denominator - 1) // lower_bound.denominator
    assert lower_ceiling == 16

    print(f"code size: {len(code)}")
    print(f"non-codeword signatures: {len(signatures)} distinct, 0 empty")
    print("signature-size distribution: 1:16 2:16 3:16")
    print(f"published lower-bound ceiling at n=6: {lower_ceiling}")
    print("verified: gamma^LD(Q_6) = 16")


if __name__ == "__main__":
    main()

File: coding_theory/q6_locating_dominating_16/verify_q6_ld16.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <iostream>
#include <map>
#include <set>
#include <vector>

int main() {
    constexpr int n = 6;
    const std::array<int, 16> listed = {
        0, 5, 11, 14, 18, 23, 25, 28,
        34, 36, 41, 47, 51, 53, 56, 62,
    };
    const std::set<int> code(listed.begin(), listed.end());

    // Independently reconstruct the algebraically specified code.
    std::set<int> generated;
    for (int x1 = 0; x1 < 2; ++x1)
        for (int x4 = 0; x4 < 2; ++x4)
            for (int x5 = 0; x5 < 2; ++x5)
                for (int x6 = 0; x6 < 2; ++x6) {
                    const int x2 = x1 ^ x4 ^ x5 ^ x6;
                    const int x3 = x1 ^ x4 ^ x6 ^ (x1 & x5) ^ (x1 & x6);
                    generated.insert((x1 << 5) | (x2 << 4) | (x3 << 3) |
                                     (x4 << 2) | (x5 << 1) | x6);
                }
    assert(generated == code);

    std::set<std::vector<int>> signatures;
    std::map<int, int> size_distribution;
    for (int vertex = 0; vertex < (1 << n); ++vertex) {
        if (code.count(vertex) != 0) continue;
        std::vector<int> signature;
        if (code.count(vertex) != 0) signature.push_back(vertex);
        for (int coordinate = 0; coordinate < n; ++coordinate) {
            const int neighbor = vertex ^ (1 << coordinate);
            if (code.count(neighbor) != 0) signature.push_back(neighbor);
        }
        std::sort(signature.begin(), signature.end());
        assert(!signature.empty());
        assert(signatures.insert(signature).second);
        ++size_distribution[static_cast<int>(signature.size())];
    }
    assert(signatures.size() == 48);
    const std::map<int, int> expected_distribution = {{1, 16}, {2, 16}, {3, 16}};
    assert(size_distribution == expected_distribution);

    const int numerator = n * n * (1 << (n + 1));
    const int denominator = n * n * n + 2 * n * n + 3 * n - 2;
    const int lower_ceiling = (numerator + denominator - 1) / denominator;
    assert(lower_ceiling == 16);

    std::cout << "code size: " << code.size() << '\n'
              << "non-codeword signatures: " << signatures.size()
              << " distinct, 0 empty\n"
              << "signature-size distribution: 1:16 2:16 3:16\n"
              << "published lower-bound ceiling at n=6: " << lower_ceiling << '\n'
              << "verified: gamma^LD(Q_6) = 16\n";
}
